Flash updates, implement Flash trait for SAM4L

bradjc opened the pull request:
The purpose of this PR is to update the SAM4L flash driver to provide a HIL trait so that capsules can use it. The `Flash` trait exists so I used that one, but made it explicitly for page-level read and writes.

I successfully used this to implement a capsule that allows apps to write their own flash (after disabling the MPU).

The vision is to end up with a stack like this:

```
                Userspace
+----------------------------------------------------+
+----------------------------------------------------+

+----------+ +----------------------------+
|          | |                            |
| AppFlash | | Nonvolatile Storage (#430) | Others...
|          | |                            |
+----------+ +----------------------------+

      +-------------------------------------------+
      hil::nonvolatile_storage::Nonvolatile_Storage
      +-------------------------------------------+

+----------------------------------------------------+
|                                                    |
|  Arbitrary Read and Write (maps unaliged access    |
|                            to page boundaries)     |
+----------------------------------------------------+

       +--------------------------------------+
       hil::flash::Flash (page le|el interface)
       +--------------------------------------+

+------------------+ +---------------------+
|                  | |                     |
| MCU Flash Driver | | External Flash Chip | Others...
|                  | |                     |
+------------------+ +---------------------+

                          hil::spi::Spi

```

alevy commented on kernel/src/hil/flash.rs:
```diff
@@ -12,14 +16,14 @@ pub trait Flash {
-    /// Write data
-    fn write(&self, offset: usize, buf: &'static mut [u8]);
+    /// Write a page of flash.
+    fn write_page(&self, page_number: usize, buf: &'static mut [u8]) -> ReturnCode;
```
Any implementation will have to truncate `buf` if it is larger than the page size, right? If so, perhaps it should be made explicit in the comment?

ppannuto replied:
I think it'd actually be better if a buffer too large failed with `ESIZE`, silent truncation on the storage pass feels not great to me

alevy replied:
Right... So ideally, then, the `buf` arguments would be reference to fixed size arrays. Or is there a problem that different flash controllers have different page sizes?

So something like:

```rust
fn write_page(&self, page_number: usize, buf: &'static mut [u8; 4096])
```

would be too restrictive for a general trait?

ppannuto replied:
Exactly. 1K pages, 16K pages, etc are reasonable

alevy replied:
So do we want something like:

```
trait Flash<B: AsMut<[u8]> {
    fn write_page(&self, page_number: usize, buf: &'static mut B);
}
```

and in the client you'd say something like:

```rust
struct MyFlashCapsule<B: AsRef<[u8], F: Flash<B>> {
    buffer: TakeCell<&'static mut B>,
    flash: &F
    ...
}

struct<B: AsRef<[u8], F: Flash<B>> MyFlashCapsule<B, F> {
    ...
    self.buf.take.map(|buf| {
            buf.as_mut()[12] = 32;
            self.flash.write_page(1, buf);
    });
    ...
}
```

Which wouldn't be guarnateed to have the right size

bradjc replied:
Does that get us something that `buf: &'static mut [u8]` doesn't?

alevy replied:
Yeah, I guess I didn't finish the example enough to illustrate why. An implementation might look like:

```rust
/// A non-existent flash hw controller with 16 byte pages
struct FlashController {
    // ... hw registers ...
}

impl Flash<[u8; 16]> for FlashController {
    ...
    fn write_page(&self, page_number: usize, &'static [u8; 16]) {
        ... // some implementation
    }
    ...
}
```

This would eliminate runtime checking because it would be impossible to pass anything except for a 16-byte array to `write_page`. In fact, technically the `AsMut` constraint isn't strictly necessary but it lets clients always convert whatever type is there into a byte-slice temporary without knowing the refified type (so I can write capsules that are generic across implementations and still do useful things).

An annoying downside I've elided is that implementations would probably have to create their own type since Rust only comes with `impl`s of `AsMut` for arrays up to like 24 bytes or something, and pages are much larger than that. It's not super hard:

```rust
pub struct MySpecificPage([u8; 4096]);

impl AsMut<[u8]> for MySpecificPage {
    fn as_mut(&self) -> &[u8] {
        &self.0
    }
}
```

And we could certainly provide implementations for common page sizes.

alevy replied:
Like, the problem with the way it's implemented right now is how the heck do I know I'm passing in something that's not going return an error because it's not exactly the right size? What do I do if it isn't?

I realize the existing HIL doesn't deal with this, but I assume(d) the assumption there was that the implementation would be able to deal with various sizes (e.g. truncate).

bradjc commented:
Updated with help from @brghena and @alevy. The major change was updating the `Flash` interface to specify the page size so that users of the `Flash` trait would be required to pass buffers to the flash layer that are the same size as a page. This is done with an associated `type Page` that wraps a u8 buffer of the correct length.

I also added many more comments to provide some guidance on how to use the `Flash` trait.

brghena commented on chips/sam4l/src/flashcalw.rs:
```diff
@@ -824,46 +854,74 @@ impl FLASHCALW {
+    }
+
+    fn read_page(&self, page_number: usize, buf: &'static mut Self::Page) -> ReturnCode {
+        self.read_range(page_number * 512, buf.len(), buf)
```
Change 512 to PAGE_SIZE

bradjc commented:
Updated to use `PAGE_SIZE` instead of a constant.
